pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,205 @@
+//
+// bpf_link (available with kernel version >= 5.7): in order for the program<->cgroup
+// association to outlive the userspace process, the link (not the program) needs to be pinned.
+// Removing the pinned link on bpffs breaks the association.
+// Cilium will only use links on fresh installs and if available in the kernel.
+// On upgrade, a link can be updated using link.Update(), which will atomically replace the
+// currently running bpf program.
+//
+// PROG_ATTACH (all kernel versions pre 5.7 that cilium supports): by definition the association
+// outlives userspace as the cgroup will hold a reference  to the attached program and detaching
+// must be done explicitly using PROG_DETACH.
+// This API is what cilium has been using prior to the 1.14 release and will continue to use if
+// bpf_link is not available.
+// On upgrade, cilium will continue to seamlessly replace old programs with the PROG_ATTACH API,
+// because updating it with a bpf_link could cause connectivity interruptions.
+
+package socketlb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"golang.org/x/sys/unix"
+)
+
+var attachTypes = map[string]ebpf.AttachType{
+	Connect4:     ebpf.AttachCGroupInet4Connect,
+	SendMsg4:     ebpf.AttachCGroupUDP4Sendmsg,
+	RecvMsg4:     ebpf.AttachCGroupUDP4Recvmsg,
+	GetPeerName4: ebpf.AttachCgroupInet4GetPeername,
+	PostBind4:    ebpf.AttachCGroupInet4PostBind,
+	PreBind4:     ebpf.AttachCGroupInet4Bind,
+	Connect6:     ebpf.AttachCGroupInet6Connect,
+	SendMsg6:     ebpf.AttachCGroupUDP6Sendmsg,
+	RecvMsg6:     ebpf.AttachCGroupUDP6Recvmsg,
+	GetPeerName6: ebpf.AttachCgroupInet6GetPeername,
+	PostBind6:    ebpf.AttachCGroupInet6PostBind,
+	PreBind6:     ebpf.AttachCGroupInet6Bind,
+}
+
+// attachCgroup attaches a program from spec with the given name to cgroupRoot.
+// If the kernel supports it, the resulting bpf_link is pinned to pinPath.
+//
+// Upgrades from prior Cilium versions will continue to be handled by a PROG_ATTACH
+// to replace an old program attached to a cgroup.
+func attachCgroup(spec *ebpf.Collection, name, cgroupRoot, pinPath string) error {
+	prog := spec.Programs[name]
+	if prog == nil {
+		return fmt.Errorf("program %s not found in ELF", name)
+	}
+
+	// A previously created link is found and updated.
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Update(prog); err != nil {
+			return fmt.Errorf("update link %s: %w", pinPath, err)
+		}
+
+		log.Debugf("updated link %s for program %s", pinPath, name)
+		return nil
+	}
+	log.Debugf("no pinned link at '%s'", pinPath)
+
+	cg, err := os.Open(cgroupRoot)
+	if err != nil {
+		return fmt.Errorf("open cgroup %s: %w", cgroupRoot, err)
+	}
+	defer cg.Close()
+
+	// no previous link is found, try to create one. This will only work on fresh installs.
+	l, err = link.AttachRawLink(link.RawLinkOptions{
```

Comment:
This could use a comment around why `RawLink` was used instead of `CgroupLink`. Timo's [comment](https://github.com/cilium/cilium/pull/23557/#issuecomment-1473879178) explains the reasoning quite well.

Reply:
I find documenting why a different API hasn't been used confusing for future readers, so I left the comment we have above the function call untouched.